### DISCUSSION TITLE
Kickr core conversion

### DISCUSSION THREAD

**Discussion by ghost** [doudar/SmartSpin2k#343] [Ideas]
Hi all,

I have a kickr core that had smoke come out of the PCB, after inspection there were a few blown caps and possible power surge, resulting in a dead microprocessor which is unrecoverable. I was about to give up and bin it when I came across this amazing project. With a few software tweaks, I reckon I can almost cut and paste this module into the kicker core. Unfortunately  I dont even know where to start with CPP. I can prob compile and upload code but any editing is beyond me. Is there anyone who could give me some pointers or place to start with the following mods:

-cadence sensor from Hall effect sensor (interrupts?)
-instead of stepper motor control, the feedback simply increases / decrease the PWM on a pin. (this drives a mosfet which increases magnetic resistance)

these two changes would be enough to get something up and running and save my kickr from the garbage. How would I go about these (hopefully) simple changes?

many thanks

**Comment by doudar** [comment]
You’d want to scale rtConfig.getCurrentIncline() to a 0-255 scale and output that on any pin not currently being used. Presumably you could do all this in main.cpp . 

Good Luck!

 I’m out on vacation until next week but possibly @eMadman or @MarkusSchneider has more insight into this. 
 
Edit CurrentIncline is what you want, not stepper position.

**Reply by ghost** [reply]
Awesome thanks, I'm still getting my head around GitHub, its been a big learning curve this weekend. Im now at a point where I can wire it up to the kicker, make changes in VS code and  push these OTA. I will have a play with the rtConfig.getStepperPosition() you mention. the kickr actually uses an optical sensor (not Hall effect) that registers 4 pulses per revolution.  If I can detect this on a pin and pass it though to the cadence rate then I'm looking pretty good. Anyway I have more than enough to keep me busy for the next while, cheers

**Comment by MarkusSchneider** [comment]
Hi @rogers2000,
are you able to do the electrical wiring and switches? You maybe have to connect an optocoupler for the cadence sensor and you may have to build a transistor amplifier for pwm out to adjust voltage/current.
For the cadence sensor is a [discussion](https://github.com/doudar/SmartSpin2k/issues/271) open requesting to connect a reed switch (maybe also a solution for you). What voltage level do you need as far as i know ESP32 uses 3.3V.

To set the PWM output you can replace this method:
https://github.com/doudar/SmartSpin2k/blob/c78614385de2f2f1ed31473909ee7274075f4d86/src/Main.cpp#L216-L277

rtConfig.getTargetIncline() gives you the requested incline from the Trainer (Zwift). If you want to use the shifter you have can also use rtConfig.getShifterPosition() * userConfig.getShiftStep()
see line 238 + 239.

You'll find the value of the targetIncline in the logs of SS2K in ERG Mode. This will help you to find the correct factors for the pwm out.
If you like, you can also use userConfig.getInclineMultiplier(). Its a correction factor to take the bike specific resistance in account. 
You can use this setting to tune the factor for pwm out.

You should also connect a power meter for ERG 😉.

**Reply by ghost** [reply]
great thanks, I can probably hack up a PWM knowing now where to insert the code. Any idea where to insert code that will give a readout of the current cadence? I would need to bypass the bluetooth code trying to connect to a sensor. Once I know where the code would go I was thinking of attaching an interrupt pin and counting the RPM through the pulses received. 

I plan to use as much of the original circuitry as possible (there really isn't much inside the kickr!), the micro controller is dead but the rest of the circuitry seems fine. It is all based around a 3.3v power rail so I can mostly plug and play with a esp32. There is an opto sensor that consists of an LED diode emitter / receiver pair that will pulse a HIGH at 3.3v onto an I/O pin, four times per revolution of the flywheel.  

The resistance circuitry is very minimal, there is the main wires from the coil inside the flywheel which essentially go straight to a mosfet which is controlled by the I/O on the micro. I haven't dealt with magnetic braking, but I assume that when the mosfet is fully open then there is no resistance, and when the mosfet is full closed then there is maximum braking through the opposing magnetic eddy currents. Anything in between (i.e. PWM) will provide a proportional braking effect. 

The best thing is that all this circuitry and associated transistor switches / amps etc are on the PCB and I can literally tap off it. I might even put an Arduino on to test my theories first and see how the braking goes before hacking up the SmartSpin code as I am more comfortable with basic Arduino functions. 

ERG through power meter is a great idea, I think the original kickr can do power measurement through some magnetic resistance feedback and cadence calculation, but just thinking about that is currently more headache than its worth!

**Comment by MarkusSchneider** [comment]
If you want to connect power meter/pedal, cadence is also sent by them via BLE. So nothing must changed.
Without power meter you must simulate power via HR monitoring. I recommend to use a power meter.

If you like to use the cadence sensor from kickr you have to set:
https://github.com/doudar/SmartSpin2k/blob/c78614385de2f2f1ed31473909ee7274075f4d86/include/SmartSpin_parameters.h#L63-L64
Should work if no cadence sensor via BLE is connected. Best way to do this is to create an interrupt function in main.cpp and use the setSimulatedCad Function

**Comment by ghost** [comment]
Update: thanks for all the help, I have managed to connect up all the kickr hardware to the esp32. I now have power and cadence through a hardware interrupt and I get get my guy moving on zwift. It is just pseudo watt measurement through a rough calculation I made up but it seems the speed and power output on swift are in the ballpark to be usable. 

Next is the PWM, but I dont think I will have any major hassles, I now have a direct I/O line that controls the braking (0v = full brake applied, 3.3v = no brake applied as tested) so thanks @MarkusSchneider for the pointers on where to start with the code. I even have the original kickr LED lights in use so I can keep it compact and professional looking. 

saved me a lot of money!

(for anyone looking at my git fork, I know there are inefficiencies in the code, especially the debounce for the cadence. Beware!)

edit: now have PWM working with feedback from the simulated incline. a simple implementation of the ledc library's enough to get it up and running

**Comment by doudar** [comment]
I looked over your fork and it’s pretty cool! Great job getting it working. This is something that I was curious if it was a possibility. Congratulations on getting it working!

**Comment by doudar** [comment]
@rogers2000 , you’ll probably want to bring your fork up to date with the latest develop as well. I’ve switched to using a more stable file system which looks like it solved quite a few issues. 

The only downside to the new file system is that upgrading the fs via OTA updates isn’t currently implemented but that’s not an issue in your use case anyway.

**Reply by ghost** [reply]
I have been away with work the last week. I will look to update the fork and then get the Kickr back in one piece with USB access for updates. Once I have the bike hooked up I  will tweak some of the code for better results. Thanks again!

**Comment by minix1234** [comment]
I'm contemplating trying this for an old Computrainer.  would need to drive a 5 volt DC MOSFET on the controls side.   

how's your effort going?

**Reply by ghost** [reply]
Hey, 
I've pretty much left it at the code that's available in my fork. I got cadence working, and I made up a power measurement (1:1 cadence I think) as it required a power measurement to get the avatar moving on zwift. The cadence seemed to work relatively well using an interrupt, I was tinkering with averaging it out over a number of cycles to get a smoother start / stop. I stopped working on it as I decided to repair my kickr to get the full ant+ functions for my Garmin watch. I was able to do this by buying a generic wifi module and extracting the firmware from the kickr core and uploading it to the new module. 

Anyway, I was happy with the progress I made on the esp32 - it read data from swift and applied the brake on the kickr through the mosfet. The PWM worked well and it felt smooth and the braking was even throughout different resistance levels. I don't think it would be too difficult to tune it to have a pretty good experience while riding using just a mosefet as you've mentioned above.

**Reply by minix1234** [reply]
glad to hear it.  

I'll have a look at you fork.  I'm thinking this could be a good alternate to the other methods of integrating Computrainers to third party apps (rouvy, SF, etc)

**Reply by doudar** [reply]
Awesome! If you get it to the point where maybe we could merge it back into the main, please let us know.

**Reply by MortenVinding** [reply]
> Hey, I've pretty much left it at the code that's available in my fork. I got cadence working, and I made up a power measurement (1:1 cadence I think) as it required a power measurement to get the avatar moving on zwift. The cadence seemed to work relatively well using an interrupt, I was tinkering with averaging it out over a number of cycles to get a smoother start / stop. I stopped working on it..............

Hi @andrewrobertdev,

I have copied your modifications to my own fork at

My bike is only sending one cadence impulse per revolution, but the RPM counter in Zwift is showing about 1/10 off.
I'm struggling to figure out your code for RPM calculations:
```
    detachInterrupt(CADENCE_PIN);
    // timeNew = millis() - timeOld;
    RPM = 60*100/(millis() - timeOld)*rpmCount;
    timeOld = millis();
    rpmCount = 0;
```
Shouldn't the calculation be 60*1000 since it is milliseconds (1/1000 second)?

Here is my fork: https://github.com/MortenVinding/SmartSpin2k-gpio
I managed to update it to the current development branch (I hope).

**Reply by doudar** [reply]
Somewhat relevant #619
